File: apps/mobile/lib/models/api_models.dart
import '../i18n/app_strings.dart';

/// Minimal exercise shape returned by list/search endpoints.
class ExerciseMinimal {
  final int id;
  final String name;

  ExerciseMinimal({required this.id, required this.name});

  factory ExerciseMinimal.fromJson(Map<String, dynamic> json) =>
      ExerciseMinimal(id: json['id'] as int, name: json['name'] as String);
}

/// Detailed exercise shape (ExerciseStandard from the API).
class ExerciseDetail {
  final int id;
  final String name;
  final List<String> primaryMuscles;
  final String? category;
  final String? force;
  final String? mechanic;
  final String? difficulty;
  final List<String> steps;

  ExerciseDetail({
    required this.id,
    required this.name,
    required this.primaryMuscles,
    this.category,
    this.force,
    this.mechanic,
    this.difficulty,
    required this.steps,
  });

  factory ExerciseDetail.fromJson(Map<String, dynamic> json) => ExerciseDetail(
        id: json['id'] as int,
        name: json['name'] as String,
        primaryMuscles: _stringList(json['primary_muscles']),
        category: json['category'] as String?,
        force: json['force'] as String?,
        mechanic: json['mechanic'] as String?,
        difficulty: json['difficulty'] as String?,
        steps: _stringList(json['steps']),
      );
}

/// Available filter values for building the picker UI.
class ExerciseFilters {
  final List<String> muscles;
  final List<String> difficulties;
  final List<String> categories;

  ExerciseFilters({
    required this.muscles,
    required this.difficulties,
    required this.categories,
  });

  factory ExerciseFilters.fromJson(Map<String, dynamic> json) =>
      ExerciseFilters(
        muscles: _stringList(json['muscles']),
        difficulties: _stringList(json['difficulties']),
        categories: _stringList(json['categories']),
      );
}

List<String> _stringList(dynamic value) =>
    (value as List?)?.map((e) => e.toString()).toList() ?? const [];

/// A training the user saved on the backend (Mongo).
class SavedTraining {
  final String id;
  final String name;
  final List<SavedExercise> exercises;
  // Provenance of the training: "manual" for user-created, "generated" for
  // plan-builder output. Drives where the card renders on the trainings list.
  final String source;
  // Last time a session was logged against this training. When non-null we
  // treat the card as "done" — sorted to the end of the generated row but
  // still visible.
  final DateTime? doneAt;
  // Optional plan grouping (only on generated trainings).
  final int? weekIndex;
  final int? dayIndex;
  // Discipline: "strength" (regular Edzések) or "hyrox" (HYROX plan). Drives
  // which tab the training belongs to.
  final String discipline;
  // HYROX 12-week phase (1–4), only set on hyrox trainings.
  final int? phase;

  SavedTraining({
    required this.id,
    required this.name,
    required this.exercises,
    this.source = 'manual',
    this.doneAt,
    this.weekIndex,
    this.dayIndex,
    this.discipline = 'strength',
    this.phase,
  });

  factory SavedTraining.fromJson(Map<String, dynamic> json) => SavedTraining(
        id: json['_id']?.toString() ?? '',
        name: json['name'] as String? ?? '',
        exercises: ((json['exercises'] as List?) ?? [])
            .map((e) => SavedExercise.fromJson(e as Map<String, dynamic>))
            .toList(),
        source: json['source'] as String? ?? 'manual',
        doneAt: _parseTrainingDate(json['doneAt']),
        weekIndex: (json['weekIndex'] as num?)?.toInt(),
        dayIndex: (json['dayIndex'] as num?)?.toInt(),
        discipline: json['discipline'] as String? ?? 'strength',
        phase: (json['phase'] as num?)?.toInt(),
      );

  int get totalSets => exercises.fold(0, (a, e) => a + e.sets.length);
  bool get isGenerated => source == 'generated';
  bool get isDone => doneAt != null;
  bool get isHyrox => discipline == 'hyrox';
}

DateTime? _parseTrainingDate(dynamic v) =>
    v is String ? DateTime.tryParse(v)?.toLocal() : null;

class SavedExercise {
  final String exerciseId;
  final String name;
  final String? category;
  final String gifUrl;
  final List<String> targetMuscles;
  final String progressionStrategy;
  // HYROX metric (null → classic strength). See lib/hyroxMetrics on the backend.
  final String? metric;
  final String? stationKey;
  final String? note;
  final List<SavedSet> sets;

  SavedExercise({
    required this.exerciseId,
    required this.name,
    this.category,
    this.gifUrl = '',
    this.targetMuscles = const [],
    this.progressionStrategy = 'linear',
    this.metric,
    this.stationKey,
    this.note,
    required this.sets,
  });

  factory SavedExercise.fromJson(Map<String, dynamic> json) => SavedExercise(
        exerciseId: json['exerciseId']?.toString() ?? '',
        name: json['name'] as String? ?? '',
        category: json['category'] as String?,
        gifUrl: json['gifUrl'] as String? ?? '',
        targetMuscles: _stringList(json['targetMuscles']),
        progressionStrategy:
            json['progressionStrategy'] as String? ?? 'linear',
        metric: json['metric'] as String?,
        stationKey: json['stationKey'] as String?,
        note: json['note'] as String?,
        sets: ((json['sets'] as List?) ?? [])
            .map((e) => SavedSet.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

/// A completed workout: when it ran and exactly what was logged.
class WorkoutSession {
  final String id;
  final String name;
  final DateTime? startedAt;
  final DateTime? finishedAt;
  final List<SavedExercise> exercises;

  WorkoutSession({
    required this.id,
    required this.name,
    required this.startedAt,
    required this.finishedAt,
    required this.exercises,
  });

  factory WorkoutSession.fromJson(Map<String, dynamic> json) => WorkoutSession(
        id: json['_id']?.toString() ?? '',
        name: json['name'] as String? ?? '',
        startedAt: _parseDate(json['startedAt']),
        finishedAt: _parseDate(json['finishedAt']),
        exercises: ((json['exercises'] as List?) ?? [])
            .map((e) => SavedExercise.fromJson(e as Map<String, dynamic>))
            .toList(),
      );

  int get totalSets => exercises.fold(0, (a, e) => a + e.sets.length);
  int get doneSets =>
      exercises.fold(0, (a, e) => a + e.sets.where((s) => s.done).length);
  Duration? get duration => (startedAt != null && finishedAt != null)
      ? finishedAt!.difference(startedAt!)
      : null;
}

DateTime? _parseDate(dynamic v) =>
    v is String ? DateTime.tryParse(v)?.toLocal() : null;

class SavedSet {
  final double kg;
  final int reps;
  final bool done;
  // HYROX metric fields (null for ordinary strength sets).
  final double? distanceM;
  final int? seconds;
  final double? targetKg;

  SavedSet({
    required this.kg,
    required this.reps,
    this.done = false,
    this.distanceM,
    this.seconds,
    this.targetKg,
  });

  factory SavedSet.fromJson(Map<String, dynamic> json) => SavedSet(
        kg: (json['kg'] as num?)?.toDouble() ?? 0,
        reps: (json['reps'] as num?)?.toInt() ?? 0,
        done: json['done'] as bool? ?? false,
        distanceM: (json['distanceM'] as num?)?.toDouble(),
        seconds: (json['seconds'] as num?)?.toInt(),
        targetKg: (json['targetKg'] as num?)?.toDouble(),
      );
}

/// A stagnating exercise: its best e1RM hasn't grown for several weeks.
class StagnationItem {
  final String exerciseId;
  final String name;
  final int weeksStagnant;
  final double e1rm;
  final List<String> tips;

  StagnationItem({
    required this.exerciseId,
    required this.name,
    required this.weeksStagnant,
    required this.e1rm,
    this.tips = const [],
  });

  factory StagnationItem.fromJson(Map<String, dynamic> json) => StagnationItem(
        exerciseId: json['exerciseId']?.toString() ?? '',
        name: json['name'] as String? ?? '',
        weeksStagnant: (json['weeksStagnant'] as num?)?.toInt() ?? 0,
        e1rm: (json['e1rm'] as num?)?.toDouble() ?? 0,
        tips: _stringList(json['tips']),
      );

  /// `Bench Press — stalled for 3 weeks. Try: deload / variation / volume drop`
  String get advice {
    final tipText = tips.isEmpty
        ? ''
        : tFmt('stagnation.try', {'tips': tips.join(' / ')});
    return tFmt('stagnation.advice', {'name': name, 'weeks': weeksStagnant}) +
        tipText;
  }
}

/// The auto-progression engine's next-set suggestion for one exercise.
class ProgressionSuggestion {
  final double suggestedKg;
  final int suggestedReps;
  final String reason;

  ProgressionSuggestion({
    required this.suggestedKg,
    required this.suggestedReps,
    this.reason = '',
  });

  factory ProgressionSuggestion.fromJson(Map<String, dynamic> json) =>
      ProgressionSuggestion(
        suggestedKg: (json['suggestedKg'] as num?)?.toDouble() ?? 0,
        suggestedReps: (json['suggestedReps'] as num?)?.toInt() ?? 0,
        reason: json['reason'] as String? ?? '',
      );

  /// `82,5 × 8` — Hungarian decimal comma, trailing `.0` trimmed.
  String get label {
    final kg = suggestedKg == suggestedKg.roundToDouble()
        ? suggestedKg.toInt().toString()
        : suggestedKg.toString().replaceAll('.', ',');
    return '$kg × $suggestedReps';
  }
}

/// XP / rank summary returned alongside a saved session. `unlocked` is true
/// when the awarded XP crossed a rank threshold — the client uses that to
/// decide whether to push the RankUpOverlay.
class RankDelta {
  final int previousRank;
  final int newRank;
  final int xpAwarded;
  final bool unlocked;

  const RankDelta({
    required this.previousRank,
    required this.newRank,
    required this.xpAwarded,
    required this.unlocked,
  });

  factory RankDelta.fromJson(Map<String, dynamic> json) => RankDelta(
        previousRank: (json['previousRank'] as num?)?.toInt() ?? 1,
        newRank: (json['newRank'] as num?)?.toInt() ?? 1,
        xpAwarded: (json['xpAwarded'] as num?)?.toInt() ?? 0,
        unlocked: json['unlocked'] as bool? ?? false,
      );
}

/// Bundle the persisted session and the rank delta so callers can react to
/// both with a single response.
class WorkoutSessionWithRank {
  final WorkoutSession session;
  final RankDelta? rankDelta;
  const WorkoutSessionWithRank({required this.session, this.rankDelta});
}

File: apps/native-mobile/src/models/apiModels.ts
// Ported 1:1 from apps/mobile/lib/models/api_models.dart. Classes keep the
// computed getters (totalSets, duration, label, advice) the Dart originals had.
import { t } from "../i18n";

function stringList(value: unknown): string[] {
  return Array.isArray(value) ? value.map((e) => String(e)) : [];
}

function parseDate(v: unknown): Date | null {
  if (typeof v !== "string") return null;
  const d = new Date(v);
  return isNaN(d.getTime()) ? null : d;
}

/** Minimal exercise shape returned by list/search endpoints. */
export class ExerciseMinimal {
  constructor(public id: number, public name: string) {}
  static fromJson(j: any): ExerciseMinimal {
    return new ExerciseMinimal(j.id as number, j.name as string);
  }
}

/** Detailed exercise shape (ExerciseStandard from the API). */
export class ExerciseDetail {
  constructor(
    public id: number,
    public name: string,
    public primaryMuscles: string[],
    public category: string | null,
    public force: string | null,
    public mechanic: string | null,
    public difficulty: string | null,
    public steps: string[],
  ) {}
  static fromJson(j: any): ExerciseDetail {
    return new ExerciseDetail(
      j.id as number,
      j.name as string,
      stringList(j.primary_muscles),
      (j.category as string) ?? null,
      (j.force as string) ?? null,
      (j.mechanic as string) ?? null,
      (j.difficulty as string) ?? null,
      stringList(j.steps),
    );
  }
}

/** Available filter values for building the picker UI. */
export class ExerciseFilters {
  constructor(
    public muscles: string[],
    public difficulties: string[],
    public categories: string[],
  ) {}
  static fromJson(j: any): ExerciseFilters {
    return new ExerciseFilters(
      stringList(j.muscles),
      stringList(j.difficulties),
      stringList(j.categories),
    );
  }
}

export class SavedSet {
  constructor(
    public kg: number,
    public reps: number,
    public done: boolean = false,
    public distanceM: number | null = null,
    public seconds: number | null = null,
    public targetKg: number | null = null,
  ) {}
  static fromJson(j: any): SavedSet {
    return new SavedSet(
      (j.kg as number) ?? 0,
      (j.reps as number) ?? 0,
      (j.done as boolean) ?? false,
      j.distanceM != null ? (j.distanceM as number) : null,
      j.seconds != null ? (j.seconds as number) : null,
      j.targetKg != null ? (j.targetKg as number) : null,
    );
  }
}

export class SavedExercise {
  constructor(
    public exerciseId: string,
    public name: string,
    public category: string | null = null,
    public gifUrl: string = "",
    public targetMuscles: string[] = [],
    public progressionStrategy: string = "linear",
    public metric: string | null = null,
    public stationKey: string | null = null,
    public note: string | null = null,
    public kcal: number | null = null,
    public sets: SavedSet[] = [],
  ) {}
  static fromJson(j: any): SavedExercise {
    return new SavedExercise(
      j.exerciseId != null ? String(j.exerciseId) : "",
      (j.name as string) ?? "",
      (j.category as string) ?? null,
      (j.gifUrl as string) ?? "",
      stringList(j.targetMuscles),
      (j.progressionStrategy as string) ?? "linear",
      (j.metric as string) ?? null,
      (j.stationKey as string) ?? null,
      (j.note as string) ?? null,
      j.kcal != null ? (j.kcal as number) : null,
      ((j.sets as any[]) ?? []).map(SavedSet.fromJson),
    );
  }
}

/** A training the user saved on the backend (Mongo). */
export class SavedTraining {
  constructor(
    public id: string,
    public name: string,
    public exercises: SavedExercise[],
    public source: string = "manual",
    public doneAt: Date | null = null,
    public weekIndex: number | null = null,
    public dayIndex: number | null = null,
    public discipline: string = "strength",
    public phase: number | null = null,
  ) {}
  static fromJson(j: any): SavedTraining {
    return new SavedTraining(
      j._id != null ? String(j._id) : "",
      (j.name as string) ?? "",
      ((j.exercises as any[]) ?? []).map(SavedExercise.fromJson),
      (j.source as string) ?? "manual",
      parseDate(j.doneAt),
      j.weekIndex != null ? (j.weekIndex as number) : null,
      j.dayIndex != null ? (j.dayIndex as number) : null,
      (j.discipline as string) ?? "strength",
      j.phase != null ? (j.phase as number) : null,
    );
  }
  get totalSets(): number {
    return this.exercises.reduce((a, e) => a + e.sets.length, 0);
  }
  get isGenerated(): boolean {
    return this.source === "generated";
  }
  get isDone(): boolean {
    return this.doneAt != null;
  }
  get isHyrox(): boolean {
    return this.discipline === "hyrox";
  }
}

/** A completed workout: when it ran and exactly what was logged. */
export class WorkoutSession {
  constructor(
    public id: string,
    public trainingId: string | null,
    public name: string,
    public startedAt: Date | null,
    public finishedAt: Date | null,
    public exercises: SavedExercise[],
  ) {}
  static fromJson(j: any): WorkoutSession {
    return new WorkoutSession(
      j._id != null ? String(j._id) : "",
      j.trainingId != null ? String(j.trainingId) : null,
      (j.name as string) ?? "",
      parseDate(j.startedAt),
      parseDate(j.finishedAt),
      ((j.exercises as any[]) ?? []).map(SavedExercise.fromJson),
    );
  }
  get totalSets(): number {
    return this.exercises.reduce((a, e) => a + e.sets.length, 0);
  }
  get doneSets(): number {
    return this.exercises.reduce(
      (a, e) => a + e.sets.filter((s) => s.done).length,
      0,
    );
  }
  /** Duration in milliseconds, or null. */
  get durationMs(): number | null {
    return this.startedAt != null && this.finishedAt != null
      ? this.finishedAt.getTime() - this.startedAt.getTime()
      : null;
  }
  get totalKcal(): number {
    return this.exercises.reduce((a, e) => a + (e.kcal ?? 0), 0);
  }
  get hasKcal(): boolean {
    return this.exercises.some((e) => e.kcal != null);
  }
}

/** A stagnating exercise: its best e1RM hasn't grown for several weeks. */
export class StagnationItem {
  constructor(
    public exerciseId: string,
    public name: string,
    public weeksStagnant: number,
    public e1rm: number,
    public tips: string[] = [],
  ) {}
  static fromJson(j: any): StagnationItem {
    return new StagnationItem(
      j.exerciseId != null ? String(j.exerciseId) : "",
      (j.name as string) ?? "",
      (j.weeksStagnant as number) ?? 0,
      (j.e1rm as number) ?? 0,
      stringList(j.tips),
    );
  }
  get advice(): string {
    const tipText =
      this.tips.length === 0 ? "" : t("stagnation.try", { tips: this.tips.join(" / ") });
    return t("stagnation.advice", { name: this.name, weeks: this.weeksStagnant }) + tipText;
  }
}

/** The auto-progression engine's next-set suggestion for one exercise. */
export class ProgressionSuggestion {
  constructor(
    public suggestedKg: number,
    public suggestedReps: number,
    public reason: string = "",
  ) {}
  static fromJson(j: any): ProgressionSuggestion {
    return new ProgressionSuggestion(
      (j.suggestedKg as number) ?? 0,
      (j.suggestedReps as number) ?? 0,
      (j.reason as string) ?? "",
    );
  }
  /** `82,5 × 8` — Hungarian decimal comma, trailing `.0` trimmed. */
  get label(): string {
    const kg =
      this.suggestedKg === Math.round(this.suggestedKg)
        ? String(Math.trunc(this.suggestedKg))
        : String(this.suggestedKg).replace(".", ",");
    return `${kg} × ${this.suggestedReps}`;
  }
}

/** XP / rank summary returned alongside a saved session. */
export class RankDelta {
  constructor(
    public previousRank: number,
    public newRank: number,
    public xpAwarded: number,
    public unlocked: boolean,
  ) {}
  static fromJson(j: any): RankDelta {
    return new RankDelta(
      (j.previousRank as number) ?? 1,
      (j.newRank as number) ?? 1,
      (j.xpAwarded as number) ?? 0,
      (j.unlocked as boolean) ?? false,
    );
  }
}

export class WorkoutSessionWithRank {
  constructor(
    public session: WorkoutSession,
    public rankDelta: RankDelta | null = null,
  ) {}
}
